runtime/doc/lua.txt

```diff
@@ -3349,4 +3349,134 @@ totable({f}, {...})                                       *vim.iter.totable()*
+                • headers table<string, string>|nil A table of key-value
+                  headers.
+                • follow_redirects boolean|nil Will follow redirects by
+                  default.
+                • data string|table|nil Data to send with the request. If a
+                  table, it will be JSON encoded. vim.net does not currently
+                  support form encoding.
+                • on_complete fun()|nil Callback function when download
+                  successfully completed.
+                • on_err fun(err: string[])|nil An optional function recieving
+                  a `stderr_buffered` string[] of curl stderr. Without
+                  providing this function, |vim.net.download()| will
+                  automatically raise an error to the user. See |on_stderr|
+                  and `stderr_buffered`.
+
+    Return: ~
+        (number) jobid A job id. See |job-control|.
+
+    See also: ~
+      • |job-control|
+      • man://curl
+
+fetch({url}, {opts})                                         *vim.net.fetch()*
+    Asynchronously make HTTP requests.
+
+    Example: >lua
+
+     -- GET a url
+     vim.net.fetch("https://example.com/api/data", {
+       on_complete = function (response)
+         -- Lets read the response!
+
+         if response.ok then
+           -- Read response text
+           local body = response.text()
+         else
+       end
+     })
+
+     -- POST to a url, sending a table as JSON and providing an authorization header
+     vim.net.fetch("https://example.com/api/data", {
+       method = "POST",
+       data = {
+         key = value
+       },
+       headers = {
+         Authorization = "Bearer " .. token
+       },
+       on_complete = function (response)
+         -- Lets read the response!
+
+         if response.ok then
+           -- Read JSON response
+           local table = response.json()
+         else
+
+         -- What went wrong?
+         vim.print(response.status)
+       end
+     })
+<
+
+    Parameters: ~
+      • {url}   (string) The request URL.
+      • {opts}  (table|nil) Optional keyword arguments:
+                • method string|nil HTTP method to use. Defaults to GET.
+                • headers table|nil A table of key-value headers.
+                • follow_redirects boolean|nil Whether to follow redirects.
+                  Defaults to true.
+                • data string|table|nil Data to send with the request. If a
+                  table, it will be JSON-encoded. vim.net does not currently
+                  support form encoding.
+                • on_complete fun(response: table)|nil Callback function when
+                  request is completed successfully. The response has the
+                  following keys:
+                  • ok boolean Whether the request was successful (status
+                    within 2XX range).
+                  • headers fun(): table<string, string[]> Function returning
```

Comment:
Up to the maintainers, but it would be nice to have `.get`, `.set`, `.append`, `.has` etc methods for Headers (for both request and response).

Otherwise always working with `table<string, string[]>` can get tedious. Appending value to existing header would require user to first check if header exists and then assign or concat to existing string. Also headers are case-insensitive, handing that in userland can be very error-prone.

Reply:
Headers can contain multiple values (cookies), hence `table<string, string[]>`. Accepting `table<string, string[]>` in opts is planned. A non-casesensitive map object to handle these operations would be useful.

Reply:
> Up to the maintainers, but it would be nice to have .get, .set, .append, .has etc methods for Headers (for both request and response).

Done